File: C#/XorShift.cs
using System;
using System.Collections.Generic;

namespace XorShifts{
	class XorShift{
		private IEnumerator<uint> r;
		public static readonly Dictionary<string,uint> defaults=new Dictionary<string,uint>(){
			{"x",123456789},
			{"y",362436069},
			{"z",521288629},
			{"w",88675123}
		};
		public readonly Dictionary<string,uint> seeds;
		public uint randCount=0;

		public XorShift(uint? _w=null,uint? _x=null,uint? _y=null,uint? _z=null){
			uint w=_w ?? (uint)Environment.TickCount;
			uint x=_x ?? w<<13;
			uint y=_y ?? (w>>9)^(x<<6);
			uint z=_z ?? y>>7;
			seeds=new Dictionary<string,uint>(){
				{"x",x},{"y",y},{"z",z},{"w",w}
			};
			r=randGen(w,x,y,z);
		}

		public IEnumerator<uint> randGen(uint w,uint x,uint y,uint z){
			uint t;
			for(;;){
				t=x^(x<<11);
				x=y;
				y=z;
				z=w;
				yield return w=(w^(w>>19))^(t^(t>>8));
			}
		}

		public uint rand(){
			randCount++;
			r.MoveNext();
			return r.Current;
		}

		public int randInt(int min=0,int max=0x7FFFFFFF){
			return (int)(rand()%(max-min+1))+min;
		}

		public float randFloat(float min=0,float max=1){
			return (float)(rand()%0xFFFF)/0xFFFF*(max-min)+min;
		}

		public T[] shuffle<T>(T[] _arr){
			var arr=(T[])_arr.Clone();
			for(int i=0;i<=arr.Length-2;i++){
				int r=randInt(i,arr.Length-1);
				T tmp=arr[i];
				arr[i]=arr[r];
				arr[r]=tmp;
			}
			return arr;
		}

		public List<T> shuffle<T>(List<T> _arr){
			var arr=new List<T>(_arr);
			for(int i=0;i<=arr.Count-2;i++){
				int r=randInt(i,arr.Count-1);
				T tmp=arr[i];
				arr[i]=arr[r];
				arr[r]=tmp;
			}
			return arr;
		}

		public class defaultSeed:XorShift{
			public defaultSeed():base(
				defaults["w"],
				defaults["x"],
				defaults["y"],
				defaults["z"]
			){}
		}
	}
}

File: C++/XorShift.cpp
#include <stdint.h>
#include <string>
#include <vector>
#include <ctime>

namespace XorShifts{
	class XorShift{
	private:
		struct dictionary{uint32_t x;uint32_t y;uint32_t z;uint32_t w;};
		uint32_t x;
		uint32_t y;
		uint32_t z;
		uint32_t w;
		uint32_t t;

	public:
		static const struct dictionary defaults;
		uint32_t randCount=0;
		struct dictionary seeds;

		XorShift(
			uint32_t w=time(nullptr),
			uint32_t x=NULL,
			uint32_t y=NULL,
			uint32_t z=NULL
		){
			x=x!=NULL? x: w<<13;
			y=y!=NULL? y: (w>>9)^(x<<6);
			z=z!=NULL? z: y>>7;
			seeds={x,y,z,w};
			this->w=w;this->x=x;this->y=y;this->z=z;
		}

		uint32_t rand(){
			randCount++;
			uint32_t t=x^(x<<11);
			x=y;
			y=z;
			z=w;
			return w=(w^(w>>19))^(t^(t>>8));
		}

		int randInt(int min=0,int max=0x7FFFFFFF){
			return rand()%(max-min+1)+min;
		}

		float randFloat(float min=0,float max=1){
			return (float)(rand()%0xFFFF)/0xFFFF*(max-min)+min;
		}

		template<typename T>
		T* shuffle(const T* _arr,T* arr,int arrLength){
			for(int i=0;i<arrLength;i++) arr[i]=_arr[i];
			for(int i=0;i<=arrLength-2;i++){
				int r=randInt(i,arrLength-1);
				T tmp=arr[i];
				arr[i]=arr[r];
				arr[r]=tmp;
			}
			return arr;
		}

		template<typename T>
		std::vector<T> shuffle(std::vector<T> arr){
			for(int i=0;i<=arr.size()-2;i++){
				int r=randInt(i,arr.size()-1);
				T tmp=arr[i];
				arr[i]=arr[r];
				arr[r]=tmp;
			}
			return arr;
		}

		static inline XorShift defaultSeed(){
			return XorShift( 
				XorShift::defaults.w,
				XorShift::defaults.x,
				XorShift::defaults.y,
				XorShift::defaults.z
			);
		}
	};

	const struct XorShift::dictionary XorShift::defaults={
		123456789,
		362436069,
		521288629,
		88675123
	};
}